### DISCUSSION TITLE
private_ident! not working? And what is the correct way to write a plugin?

### DISCUSSION THREAD

**Discussion by Jack-Works** [swc-project/swc#4848] [Q&A] [answered]
Hello! I'm new to swc and I want to write an swc plugin:

I found `private_ident!` does not generate a unique name in the output, I guess I did something wrong...

And I also find the testing process is not convenient (I tried `test_fixture`, but it looks like I need to manually write new snapshots to the filesystem so I'm not using it), did I miss something?

I also found the plugin created by `swc cli` is missing many important dependencies, I have to add them to the Cargo.toml manually including `swc_common`, `swc_ecma_transforms_testing`, `swc_ecma_parser`, `testing`

```rust
use swc_plugin::{ast::*, plugin_transform, utils::private_ident, TransformPluginProgramMetadata};

pub struct TransformVisitor {
    uses_import_meta: bool,
}

impl TransformVisitor {
    pub fn new() -> Self {
        Self {
            uses_import_meta: false,
        }
    }
}

impl Fold for TransformVisitor {
    // https://rustdoc.swc.rs/swc_ecma_visit/trait.Fold.html
    fn fold_expr(&mut self, n: Expr) -> Expr {
        match n {
            Expr::MetaProp(meta) if meta.kind == MetaPropKind::ImportMeta => {
                self.uses_import_meta = true;
                Expr::Ident(private_ident!("import_meta"))
            }
            _ => n.fold_children_with(self),
        }
    }
}

impl VisitMut for TransformVisitor {
    // https://rustdoc.swc.rs/swc_ecma_visit/trait.VisitMut.html
}

#[plugin_transform]
pub fn process_transform(program: Program, _metadata: TransformPluginProgramMetadata) -> Program {
    program.fold_with(&mut as_folder(TransformVisitor::new()))
}

#[cfg(test)]
mod tests {
    use std::fs::{read_to_string, write};
    use std::{path::PathBuf, rc::Rc};
    use swc_common::comments::SingleThreadedComments;
    use swc_ecma_parser::{Syntax};
    use swc_ecma_transforms_testing::{test, Tester};

    use crate::TransformVisitor;

    #[testing::fixture("tests/fixtures/**/input.js")]
    fn test(input: PathBuf) {
        let output = input.with_file_name("output.js");

        Tester::run(|tester| {
            let actual = tester.apply_transform(
                TransformVisitor::new(),
                "output.js",
                Syntax::Es(Default::default()),
                read_to_string(&input).unwrap().as_str(),
            )?;
            let result = tester.print(&actual, &Rc::new(SingleThreadedComments::default()));
            write(&output, result).unwrap();
            Ok(())
        })
    }
}
```

**Comment by kwonoj** [comment]
1. I think it's related to https://github.com/swc-project/swc/issues/4803 to support unresolved_mark.

2. It is intended to have minimal dependencies for the generated project via swc cli. If needed, additional dependencies can be added manually. If it needs to import swc_common however, it may need to be reexported from swc_plugin.

**Comment by daflyinbed** [comment]
If you want to use `Fold`, then you don't need `impl VisitMut for TransformVisitor` and `as_folder`

**Comment by kdy1** [comment, marked as answer]
`hygiene` pass will rename identifiers created `private_ident` **iff** there's a conflict.

**Reply by Jack-Works** [reply]
Looks like I need to use `unresolved_mark` (not for issue `private_ident` but for other cases). Is it impossible to get one before #4803 fixes?
